gem/refactorise.py

```diff
@@ -112,6 +116,175 @@ def product(*args):
             result.add(sum_indices, atomics, rest)
         return result
 
+    def argument_indices_extent(self, factor):
+        """
+        Returns the product of extents of argument indices of :param: factor
+        """
+        if self.argument_indices is None:
+            raise AssertionError("argument_indices property not initialised.")
+        return numpy.product([i.extent for i in set(factor.free_indices).intersection(self.argument_indices)])
```

Comment:
I think if you refactor to make all these new methods free functions, you will find you don't have to back-door extra properties onto the `MonomialSum`.  IOW, it would be much clearer if this were just:

```python
def index_extent(factor, argument_indices):
    """Compute the product of the indices of factor that appear in argument indices"""
    return numpy.product([i.extent for i in set(...)])
```